src/lib/address_bitset.cc

```diff
@@ -0,0 +1,70 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present
+// Datadog, Inc.
+#include "address_bitset.hpp"
+
+#include <algorithm>
+#include <functional>
+#include <unlikely.hpp>
+
+namespace ddprof {
+
+void AddressBitset::init(unsigned max_addresses) {
+  // Due to memory alignment, on 64 bits we can assume that the first 4
+  // bits can be ignored
+  _lower_bits_ignored = _k_max_bits_ignored;
+  if (_address_bitset) {
+    _address_bitset.reset();
+  }
+  _nb_bits = max_addresses;
+  _k_nb_elements = (_nb_bits) / (_nb_bits_per_elt);
+  if (_nb_bits) {
+    _nb_bits_mask = _nb_bits - 1;
+    _address_bitset = std::make_unique<std::atomic<uint64_t>[]>(_k_nb_elements);
+  }
+}
+
+bool AddressBitset::set(uintptr_t addr) {
+  uint32_t significant_bits = remove_lower_bits(addr);
+  // As per nsavoire's comment, it is better to use separate operators
+  // than to use the div instruction which generates an extra function call
+  // Also, the usage of a power of two value allows for bit operations
+  unsigned index_array = significant_bits / 64;
+  unsigned bit_offset = significant_bits % 64;
+  uint64_t bit_in_element = (1UL << bit_offset);
+  // there is a possible race between checking the value
+  // and setting it
+  if (!(_address_bitset[index_array].fetch_or(bit_in_element) &
+        bit_in_element)) {
+    // check that the element was not already set
+    ++_nb_addresses;
+    return true;
+  }
+  // Collision, element was already set
+  return false;
+}
+
+bool AddressBitset::unset(uintptr_t addr) {
+  uint64_t hash_addr = remove_lower_bits(addr);
+  int significant_bits = hash_addr & _nb_bits_mask;
```

Comment:
masking is already done in `remove_lower_bits` (contrary to what the name suggests ;) )